rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,626 @@
+
+### Implementation
+
+The bot will be implemented as a standalone tool (ie under `tool/tbot`) and
+_not_ as a service that runs in a `teleport` process.
+
+The bulk of the new logic (parsing configuration, writing certificates to a
+destination, etc.) will be implemented in a new `renew` package, which will
+enable us to start with a standalone bot, and incorporate this functionality
+into other parts of Teleport (ie database access) in the future.
+
+#### API Objects
+
+On the backend, registering a new bot (via `tctl`) creates a set of related API
+objects:
+
+- an object representing the bot itself
+- a non-interactive User for the bot to act as (when user certs are requested)
+- a role for the bot, which will contain the minimum set of permissions
+  necessary to watch for CA rotations, as well as the ability to impersonate
+  other roles
+- a token that the bot can use to join the cluster for the first time
+
+In order to discover related objects and ensure that no "zombie" items are left behind,
+`tctl` will ensure that:
+
+- the objects related to a bot all follow a standardized naming convention that
+  includes the name of the bot
+- a new `teleport.dev/owned-by` label is applied to all resources related to a
+  particular bot
+
+Note: end users should never need to manually create, modify, or delete these
+objects. All operations on bots and their related objects is performed through
+the `tctl bots` subcommand.
+
+#### `tbot` configuration
+
+The goal is to keep tbot configuration to a minimum to make for a quick getting
+started experience, however we recognize many use cases may quickly become
+difficult to encode on the command line.
+
+To that end, we'd like to follow the philosophy of "keep simple things simple
+but make complex things possible": the simplest / most popular use cases should
+ideally require minimal (zero?) configuration beyond a concise CLI, but more
+complicated use cases can make use of a `tbot.yaml` configuration file.
+
+An example `tbot.yaml` with all advanced options specified:
+
+```yaml
+# Storage for the bot's private data
+storage:
+  # can also support any backend supporting both store + fetch
+  directory: /var/lib/tbot
+
+sinks:
+  - directory: /home/alice/.tbot
+    roles: [a, b, c]
+
+  # TODO: probably need to specify permissions (this will be ugly)
+  - directory: /home/bob/.tbot
+    roles: [b, c]
+
+  # k8s secrets support store and fetch
+  - kubernetesSecret:
+      kubeconfig: /home/user/kubeconfig
+      namespace: example
+      secretName: example-secret
+    roles: [c]
+
+  # Webhooks support only store
+  - webhook: http://localhost:8001
+    roles: [a]
+```
+
+#### External Configuration Assist
+
+In order to make it easy to consume certificates generated by `tbot`, the agent
+will have a `tbot config` subcommand that can render configuration snippets
+for common integration points.
+
+For example:
+
+```
+$ tbot start ...
+$ tbot config --ssh
```

Comment:
most likely this will be many preset commands, so subcommand will make more sense. There will be one for client and another for server, for example. The advantage of the subcommand is that each subcommand can have separate parameter flags that will let users tweak base folder for example.

Reply:
It seems there's 2 sections on the configuration assistant, so I'll merge the relevant details into this section and remove the other one.

Anyway, I also prefer the subcommand approach, it will more closely match `tsh config` too.